go/libkb/dbcleaner.go

```diff
@@ -0,0 +1,319 @@
+	}
+	return c
+}
+
+func (c *levelDbCleaner) Stop() {
+	c.log(context.TODO(), "Stop")
+	c.Lock()
+	defer c.Unlock()
+	if c.stopCh != nil {
+		close(c.stopCh)
+		c.stopCh = make(chan struct{})
+	}
+}
+
+func (c *levelDbCleaner) monitorAppState() {
+	ctx := context.Background()
+	c.log(ctx, "monitorAppState")
+	state := keybase1.AppState_FOREGROUND
+	for {
+		select {
+		case state = <-c.G().AppState.NextUpdate(&state):
+			switch state {
+			case keybase1.AppState_BACKGROUNDACTIVE:
+				c.log(ctx, "monitorAppState: attempting clean")
+				c.clean(ctx, false)
+			default:
+				c.log(ctx, "monitorAppState: attempting cancel, state: %v", state)
+				c.Lock()
+				if c.cancelCh != nil {
+					close(c.cancelCh)
+					c.cancelCh = make(chan struct{})
+				}
+				c.Unlock()
+			}
+		case <-c.stopCh:
+			c.log(context.TODO(), "monitorAppState: stop")
+			return
+		}
+	}
+}
+
+func (c *levelDbCleaner) log(ctx context.Context, format string, args ...interface{}) {
+	c.G().Log.CDebugf(ctx, fmt.Sprintf("levelDbCleaner(%s): %s", c.dbName, format), args...)
+}
+
+func (c *levelDbCleaner) setDb(db *leveldb.DB) {
+	c.Lock()
+	defer c.Unlock()
+	c.db = db
+}
+
+func (c *levelDbCleaner) cacheKey(key []byte) string {
+	return string(key)
+}
+
+func (c *levelDbCleaner) clearCache() {
+	c.cache.Purge()
+}
+
+func (c *levelDbCleaner) shouldCleanLocked(force bool) bool {
+	if c.running {
+		return false
+	}
+	if force {
+		return true
+	}
+	validCache := c.cache.Len() >= c.config.MinCacheSize
+	if c.isMobile {
+		return validCache && c.G().AppState.State() == keybase1.AppState_BACKGROUNDACTIVE
+	}
+	return validCache &&
+		c.G().GetClock().Now().Sub(c.lastRun) >= c.config.CleanInterval
+}
+
+func (c *levelDbCleaner) getDbSize(ctx context.Context) (size uint64, err error) {
+	if c.db == nil {
+		return 0, nil
+	}
+	// get the size from the start of the kv table to the beginning of the perm
+	// table since that is all we can clean
+	dbRange := util.Range{Start: tablePrefix(levelDbTableKv), Limit: tablePrefix(levelDbTablePerm)}
+	sizes, err := c.db.SizeOf([]util.Range{dbRange})
+	if err != nil {
+		return 0, err
+	}
+	return uint64(sizes.Sum()), nil
+}
+
+func (c *levelDbCleaner) clean(ctx context.Context, force bool) (err error) {
+	c.Lock()
+	// get out without spamming the logs
+	if !c.shouldCleanLocked(force) {
+		c.Unlock()
+		return nil
+	}
+	c.running = true
+	key := c.lastKey
+	c.Unlock()
+
+	defer c.G().CTraceTimed(ctx, fmt.Sprintf("levelDbCleaner(%s) clean, config: %v", c.dbName, c.config), func() error { return err })()
+	dbSize, err := c.getDbSize(ctx)
+	if err != nil {
+		return err
+	}
+
+	c.log(ctx, "dbSize: %v, cacheSize: %v",
+		humanize.Bytes(dbSize), c.cache.Len())
+	// check db size, abort if small enough
+	if !force && dbSize < c.config.MaxSize {
+		return nil
+	}
+
+	var totalNumPurged, numPurged int
+	for i := 0; i < 100; i++ {
+		select {
+		case <-c.cancelCh:
+			c.log(ctx, "aborting clean, %d runs, canceled", i)
+			break
```

Comment:
I don't think these `break` statements are doing what you want.  You would need a label to break out of the `for` loop here.

see the 3rd example here:

https://golang.org/doc/effective_go.html#switch

Reply:
ah great catch